src/decoder.ts

```diff
@@ -159,41 +154,40 @@ export const allOfDecoders = <
 ): Decoder<R> =>
   decoder((a: unknown) => {
     return decoders.reduce(
-      (result: Result<R>, decoderNext: Decoder<unknown>) => {
-        switch (result.type) {
-          case OK:
-            return decoderNext.decode(result.value) as Result<R>;
-          default:
-            return err<R>(result.message);
-        }
-      },
+      (result: Result<R>, decoderNext: Decoder<unknown>) =>
+        result.type === OK
+          ? (decoderNext.decode(result.value) as Result<R>)
+          : err<R>(result.message),
       ok<R>(a as R)
     );
   });
 
 export const exactDecoder = <T>(value: T): Decoder<T> =>
-  decoder((a: unknown) =>
-    a === value ? ok(value) : err(`not exactly ${value}`)
-  );
+  Number.isNaN(value as any)
+    ? decoder((a: unknown) =>
+        Number.isNaN(a as any) ? ok(value) : err(`not exactly ${value}`)
+      )
+    : decoder((a: unknown) =>
+        a === value ? ok(value) : err(`not exactly ${value}`)
+      );
 
 export const objectDecoder = <T>(decoderMap: DecoderMap<T>): Decoder<T> =>
   decoder((a: unknown) => {
-    if (typeof a === "object") {
-      const keys = Object.keys(decoderMap) as (keyof T)[];
-      const res: Partial<T> = {};
-      for (const key of keys) {
-        const fieldResult = decoderMap[key].decode(((a as unknown) as T)[key]);
-        switch (fieldResult.type) {
-          case OK: {
-            res[key] = fieldResult.value;
-            continue;
-          }
-          case ERR:
-            return err(`${key}: ${fieldResult.message}`);
-        }
-      }
-      return ok(res as T);
-    } else return err(`expected object, got ${typeof a}`);
+    if (typeof a !== "object") return err(`expected object, got ${typeof a}`);
+    if (!a) return err(`expect object, got ${a}`);
+
+    const keys = Object.keys(decoderMap) as (keyof T)[];
+    const res: Partial<T> = {};
+    for (const key of keys) {
+      const fieldResult = decoderMap[key].decode(((a as unknown) as T)[key]);
```

Comment:
not sure i understand the purpose of this,
poor man's pattern matching is doing its job!

Reply:
There is no difference in behavior

Reply:
exactly, but former adds readability